### DISCUSSION TITLE
Is using a `TypeAliasType` to simply forward a `TypeVar` supposed to be approved?

### DISCUSSION THREAD

**Discussion by cainmagi** [microsoft/pyright#5526] [Q&A] [answered]
Hello! I am trying to define an annotated proxy class. Previously, the type checker did not approve all my run-time compatible tries. But recently, I found the following usage will satisfy my requirement without raising a type checker error:

```python
from typing import TypeVar, Sequence
from typing_extensions import TypeAliasType, reveal_type

T = TypeVar("T")

proxy = TypeAliasType("proxy", T, type_params=(T,))


# No errors are raised, and I am happy to see that!
def test_func(val1: Sequence[proxy[int]], val2: Sequence[proxy[str]]) -> Sequence[str]:
    return val2


reveal_type(test_func)  # (val1: Sequence[int], val2: Sequence[str]) -> Sequence[str]
```

How `test_func` can accept arguments like `val1=(1, 2,)` without raising any errors. This is what I expect to get.

However, the usage of `TypeAliasType` here may be problematic. In most cases, `TypeAliasType` is not used like this but is used for providing abbr. of complicated compositional types. **I am not sure whether defining an alias with a single `TypeVar` should be approved.** If it can be approved, this will be my final solution for a "type proxy".

### **VS Code extension or command-line**

* **Pylance** version: 2023.7.20
* **Pyright** version: 74d8f3c4

**Comment by erictraut** [comment, marked as answer]
Yes, this should work fine. In fact, you shouldn't need to use `typing.TypeAliasType`. The older-style `typing.TypeAlias` should work fine here. There's currently a bug in pyright where `proxy: TypeAlias = T` does not work correctly. I'll fix this for the next release.

```python
proxy: TypeAlias = T
```

**Reply by erictraut** [reply]
Note that mypy and other type checkers have not yet added support for `typing.TypeAliasType`. It was added as part of PEP 695 in python 3.12 (which is still in beta).

**Reply by cainmagi** [reply]
Thank you! That helps me a lot. This is what I exactly want to do in the previously rejected issue #5489. I am happy to finally find a legal way to do that.

Besides, I found that `TypeAliasType` can do something that `TypeAlias` may not be able to do, for example:
```python
from typing import Sequence
from typing_extensions import TypeAliasType, TypeVar, reveal_type


T = TypeVar("T")
proxy = TypeAliasType("proxy", int, type_params=(T,))


def test_func(val1: Sequence[proxy[int]], val2: Sequence[proxy[str]]) -> Sequence[str]:
    ...


reveal_type(test_func)  # (val1: Sequence[int], val2: Sequence[int]) -> Sequence[str]
```
This usage is very weird, where `T` is not actually used, thus leading `proxy[Any]` to be expanded as `int` by `reveal_type(...)`. Currently, this usage is not treated as an error by `pyright`, should it be forbidden?

Thank you!

**Reply by erictraut** [reply]
> should it be forbidden? 

This is a bit strange, but it's valid. As you said, it's odd to define a type parameter that isn't used in the type alias.

In python 3.12, you can define this using the new `type` syntax:

```python
type proxy[T] = int
```

**Reply by cainmagi** [reply]
Thank you! If so, I think I can use this usage to provide a solution to
https://github.com/python/mypy/issues/15107

(Sorry, I misreferred another issue, I have changed it just now)

**Reply by cainmagi** [reply]
I just checked another thing:

```python
from typing import TypeVar
from typing_extensions import TypeAlias


T = TypeVar("T")
proxy: TypeAlias = T
proxy[int]
```

The above codes raise an error in run time:

#### python3.8
```log
Traceback (most recent call last):
  File "alpha-proxy.py", line 7, in <module>
    proxy[int]
TypeError: 'TypeVar' object is not subscriptable
```
#### python3.11
```log
Traceback (most recent call last):
  File "C:\Users\Yuchen Jin\Documents\GitHRC\dash-ait-pvt-analysis\alpha-proxy.py", line 7, in <module>
    proxy[int]
    ~~~~~^^^^^
TypeError: 'TypeVar' object is not subscriptable
```

It seems that `TypeAlias` annotation behaves differently from defining `TypeAliasType` when it is used to annotate `T`. Maybe it should not be treated in the same way of `TypeAliasType`?

**Reply by erictraut** [reply]
Hmm, I can't explain why the runtime would reject that. Perhaps @JelleZijlstra could shed some light on it. I'm not aware of any documentation that indicates this should not be permitted.

One potential workaround is to define the `proxy` type alias as a union between `T` and `typing.Never`.

```python
proxy: TypeAlias = T | Never
```

It's a bit of a hack, but this appears to work fine at runtime and type checks with both mypy and pyright.

**Reply by cainmagi** [reply]
Thank you! That's magic!

**Reply by hmc-cs-mdrissi** [reply]
There’s special name for generic type where type argument is unused, https://wiki.haskell.org/Phantom_type.  I haven’t seen them before in python but some languages do interesting type level stuff with them.
